programs/vote/src/vote_state/mod.rs

```diff
@@ -1925,526 +1915,78 @@ mod tests {
-                ],
-                Some(6),
-                6,
-            ),
+        // Create a new Votestate
+        let mut vote_state = VoteState::new(&VoteInit::default(), &Clock::default());
+
+        // Test data: a sequence of groups of votes to simulate having been cast, after each group a vote
+        // state update is compared to "normal" vote processing to ensure that credits are earned equally
+        let test_vote_groups: Vec<Vec<Slot>> = vec![
+            // Initial set of votes that don't dequeue any slots, so no credits earned
+            vec![1, 2, 3, 4, 5, 6, 7, 8],
+            vec![
+                9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29,
+                30, 31,
+            ],
+            // Now a single vote which should result in the first root and first credit earned
+            vec![32],
+            // Now another vote, should earn one credit
+            vec![33],
+            // Two votes in sequence
+            vec![34, 35],
+            // 3 votes in sequence
+            vec![36, 37, 38],
+            // 30 votes in sequence
+            vec![
+                39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59,
+                60, 61, 62, 63, 64, 65, 66, 67, 68,
+            ],
+            // 31 votes in sequence
+            vec![
+                69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89,
+                90, 91, 92, 93, 94, 95, 96, 97, 98, 99,
+            ],
+            // Votes with expiry
+            vec![100, 101, 106, 107, 112, 116, 120, 121, 122, 124],
+            // More votes with expiry of a large number of votes
+            vec![200, 201],
+            vec![
+                202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217,
+                218, 219, 220, 221, 222, 223, 224, 225, 226,
+            ],
+            vec![227, 228, 229, 230, 231, 232, 233, 234, 235, 236],
         ];
 
         let mut feature_set = FeatureSet::default();
         feature_set.activate(&feature_set::vote_state_update_credit_per_dequeue::id(), 1);
 
-        for data in &test_data {
-            let vote_state_votes = make_lockouts(&data.0);
-            let vote_state_root = data.1;
-            let vote_state_update_votes = make_lockouts(&data.2);
-            let vote_state_update_root = data.3;
-            let per_slot_expected_credits = data.4;
-
-            let mut vote_state = VoteState::default();
+        for vote_group in test_vote_groups {
+            // Duplicate vote_state so that the new votes can be applied one by one
+            let mut vote_state_after_votes = vote_state.clone();
 
-            vote_state.votes = vote_state_votes.clone();
-
-            vote_state.root_slot = vote_state_root;
+            vote_group
+                .iter()
+                .for_each(|s| vote_state_after_votes.process_next_vote_slot(*s, 0));
```

Comment:
`process_next_vote_slot` is kind of a hack that skips the call to `check_slots_are_valid`. 

The more safe equivalent would be `process_vote_unchecked`